File: src/NodeParser/RestTypeNodeParser.ts
import ts from "typescript";
import type { Context, NodeParser } from "../NodeParser.js";
import type { SubNodeParser } from "../SubNodeParser.js";
import type { ArrayType } from "../Type/ArrayType.js";
import type { BaseType } from "../Type/BaseType.js";
import type { InferType } from "../Type/InferType.js";
import { RestType } from "../Type/RestType.js";
import type { TupleType } from "../Type/TupleType.js";

export class RestTypeNodeParser implements SubNodeParser {
    public constructor(protected childNodeParser: NodeParser) {}
    public supportsNode(node: ts.RestTypeNode): boolean {
        return node.kind === ts.SyntaxKind.RestType;
    }
    public createType(node: ts.RestTypeNode, context: Context): BaseType {
        return new RestType(this.childNodeParser.createType(node.type, context) as ArrayType | InferType | TupleType);
    }
}

File: internal/parser/rest_type_node_parser.go
package parser

import (
	"github.com/microsoft/typescript-go/shim/ast"

	"github.com/vega/ts-json-schema-generator/internal/types"
)

// RestTypeNodeParser handles rest tuple elements like `[...string[]]`
// (src/NodeParser/RestTypeNodeParser.ts).
type RestTypeNodeParser struct {
	childNodeParser NodeParser
}

func NewRestTypeNodeParser(childNodeParser NodeParser) *RestTypeNodeParser {
	return &RestTypeNodeParser{childNodeParser: childNodeParser}
}

func (p *RestTypeNodeParser) SupportsNode(node *ast.Node) bool {
	return node.Kind == ast.KindRestType
}

func (p *RestTypeNodeParser) CreateType(node *ast.Node, ctx *Context, _ *types.ReferenceType) types.Type {
	return &types.RestType{Type: p.childNodeParser.CreateType(node.AsRestTypeNode().Type, ctx, nil)}
}
